x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -42,17 +60,175 @@
+        ensureOpen();
+        return fileInfo(name).length();
+    }
+
+    @Override
+    public Set<String> getPendingDeletions() {
+        throw unsupportedException();
+    }
+
+    @Override
+    public void sync(Collection<String> names) {
+        throw unsupportedException();
+    }
+
+    @Override
+    public void syncMetaData() {
+        throw unsupportedException();
+    }
+
+    @Override
+    public void deleteFile(String name) {
+        throw unsupportedException();
+    }
+
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) {
+        throw unsupportedException();
+    }
+
+    @Override
+    public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) {
+        throw unsupportedException();
+    }
+
+    @Override
+    public void rename(String source, String dest) {
+        throw unsupportedException();
+    }
+
+    private static UnsupportedOperationException unsupportedException() {
+        assert false : "this operation is not supported and should have not be called";
+        return new UnsupportedOperationException("Searchable snapshot directory does not support this operation");
+    }
+
+    @Override
+    public final void close() {
+        if (closed.compareAndSet(false, true)) {
+            isOpen = false;
+            // Ideally we could let the cache evict/remove cached files by itself after the
+            // directory has been closed.
+            clearCache();
+        }
+    }
+
+    public void clearCache() {
+        cacheService.removeFromCache(cacheKey -> cacheKey.belongsTo(snapshotId, indexId, shardId));
+    }
+
+    protected IndexInputStats createIndexInputStats(final long fileLength) {
+        return new IndexInputStats(fileLength);
+    }
+
+    public CacheKey createCacheKey(String fileName) {
+        return new CacheKey(snapshotId, indexId, shardId, fileName);
+    }
+
+    public CacheFile getCacheFile(CacheKey cacheKey, long fileLength) throws Exception {
+        return cacheService.get(cacheKey, fileLength, cacheDir);
     }
 
     @Override
     public IndexInput openInput(final String name, final IOContext context) throws IOException {
         ensureOpen();
-        return new SearchableSnapshotIndexInput(blobContainer, fileInfo(name), context, blobContainer.readBlobPreferredLength(),
-            BufferedIndexInput.BUFFER_SIZE);
+        final BlobStoreIndexShardSnapshot.FileInfo fileInfo = fileInfo(name);
+        final IndexInputStats inputStats = stats.computeIfAbsent(name, n -> createIndexInputStats(fileInfo.length()));
+        if (useCache) {
+            return new CacheBufferedIndexInput(this, fileInfo, context, inputStats);
+        } else {
+            long preferredLength = blobContainer.readBlobPreferredLength();
+            return new SearchableSnapshotIndexInput(blobContainer, fileInfo, context, preferredLength, BufferedIndexInput.BUFFER_SIZE);
```

Comment:
I'd like to rename this one to DirectBufferedIndexInput and to clean it up a bit in a follow up PR